### ISSUE TITLE
Issues with multiple forms

### ISSUE THREAD

**Issue by greg0ire** [schmittjoh/JMSPaymentCoreBundle#211]
I have a page with several forms. Recently, I updated this bundle and now, it looks like the second form overwrites the options of the transformer of the first one. It could very well be the case, because a change has been introduced in 473c49f1f07215e11c02dc524f571f37e16b70fa to create this transformer, which means if you have two forms, the transformer is the same : `payment.form.choose_payment_method_transformer`. Before that, since there were 2 instances of the forms, the options were kept separate.

So to sum up the issue, services should be stateless, but `ChoosePaymentMethodTransformer` isn't since it has an $options. A possible solution would be to clone it (in the setter).

**Comment by regularjack** [comment]
Thanks for reporting this. The commit you mention indeed introduces a change in behaviour. I need to think further about this to figure out the best way to fix the issue.

Just so that I understand the use case, you have multiple instances of `ChoosePaymentMethodType`? May I ask why? It never occurred to me that this could happen out there.

**Comment by greg0ire** [comment]
It's a checkout page, and there are a main payment form, and an alternative payment form. The code is very ugly, and might be stupid, because once you pick one of the payment forms, you get on a page where you only see one of them. I guess there must still be two forms in the action behind that although there is only one displayed. The same action has both forms, not 100% sure why the previous developers decided to do that.

**Comment by greg0ire** [comment]
Erratum : there is only one data transformer instanciated, but it seems that the form type is also only used once. This means cloning won't work.

**Comment by greg0ire** [comment]
I understand, the page where there is the payment choice and the page where you fill the form are actually the same. It's a JS illusion.

**Comment by greg0ire** [comment]
I found a workaround, and I think the bundle is very misused in the codebase I'm touching at. Closing.
